### PR TITLE
Add support for linear algebra ops on GPU using Cusolver:

### LINKED ISSUE #723: Support for GPU SVD
For doing physics simulation (especially quantum physics) complex SVDs are done very often. It would be a big win for our team if we could get acceleration of this on GPU.

### LINKED ISSUE #825: jax.numpy.linalg ops missing GPU implementation
I often hit messages like 
```"Singular value decomposition is only implemented on the CPU backend"```
(https://github.com/google/jax/blob/27746b8c73f9ca9928da5da40b7382ae648a5f8d/jax/lax_linalg.py)
for many `jax.numpy.linalg` ops. Examples I've hit so far are when calling:
`jax.numpy.linalg.svd`;
`jax.numpy.linalg.eigh`.

It would be nice to be able to run on GPU.

Thanks!

### PR DESCRIPTION
**Pull request by hawkinsp**
* LU decomposition
* Symmetric (Hermitian) eigendecomposition
* Singular value decomposition.

Make LU decomposition tests less sensitive to the exact decomposition; check that we have a decomposition, not precisely the same one scipy returns.

Fixes #723 
Fixes #825

### CODE REVIEW

File: jax/lax_linalg.py [hunk 1]
```diff
@@ -81,6 +83,11 @@ def _T(x): return np.swapaxes(x, -1, -2)
 def _H(x): return np.conj(_T(x))
 def symmetrize(x): return (x + _H(x)) / 2
 
+def _unpack_tuple(f, n):
```
Comment (mattjj) [inline jax/lax_linalg.py]: Nice, this leads to simplifications everywhere! There's a related function in xla.py called `xla_destructure` but it's not quite the same in that (1) it uses `len(c.GetShape(tuple_op).tuple_shapes())` to compute `n` and (2) it's not a decorator / higher-order function. I'm not suggesting reusing that function, but I wonder if it would make sense to use technique (1) here to avoid passing in `n`.

File: jaxlib/lapack.pyx [hunk 2]
```diff
@@ -358,8 +358,10 @@ def jax_getrf(c, a):
             batch_dims + (m, n),
             (num_bd, num_bd + 1) + tuple(range(num_bd - 1, -1, -1))),
       ))
+  return tuple(c.GetTupleElement(out, i) for i in range(3))
```
Comment (mattjj) [inline jaxlib/lapack.pyx]: Could use `xla_destructure(c, out)` from `jax.interpreters.xla` here, not that it saves much typing.

**Review by mattjj — APPROVED** [review]
This is amazing :') A huge lift that will give JAX users fast GPU linear algebra. This opens up a lot of new applications.

It's doubly amazing because it also serves as a set of examples for how to plug custom GPU kernels into JAX while still allowing end-to-end XLA compilation. Woo!

I left some minor comments mostly to show that I actually read the code here. My only question is: what about potrf? I want to run a giant Kalman smoother on my GPU :)

File: jax/lax_linalg.py (hunk 1 above)
Comment (hawkinsp) [reply jax/lax_linalg.py]: I'd like to entirely delete this function as soon as we have raised the minimum Jaxlib version, which I will probably do as soon as we have built new wheels. So I'm tempted to leave it as is for now.

File: jaxlib/lapack.pyx (hunk 2 above)
Comment (hawkinsp) [reply jaxlib/lapack.pyx]: Probably better to avoid jaxlib -> jax dependencies, though...

**Comment by hawkinsp** [comment]
The reason that `potrf` (Cholesky decomposition) is missing is because it already has GPU support. I added it as a first-class op to XLA. Now it's there I'm probably going to leave it that way. The custom-call approach is a much easier way to add more ops, however.

**Comment by clemisch** [comment]
This is great! Does this also include batched Eigendecomposition / SVD on GPU?

**Comment by hawkinsp** [comment]
It includes batched symmetric eigendecomposition (although that's implemented by performing one eigendecomposition per batch element).

The kernel added here has support for batched SVD internally, but it turns out I never got around to adding batched SVD support to CPU either, so some of the Python code needs updating to understand batching. My plan is to add batched SVD in another PR. Do you have a particular use for batched SVD? If you do, it would be interesting to know what shapes/sizes you need given there's a choice of implementations on GPU (Jacobi vs QR).

**Comment by Thenerdstation** [comment]
This is awesome! Thank you so much Peter!

**Comment by clemisch** [comment]
Thanks @hawkinsp for your response! 

My particular use for batched SVD would be a regularizer for multi-channel CT reconstruction (from this [paper](https://www.ncbi.nlm.nih.gov/pubmed/25658985)). I have implemented this by hand, including gradient, but it would be cool to have support in JAX. 

---

If you're still interested, my common scenario is an array with shape `(3, 3, 300, 300, 300)`. I would compute the SVD over the first two axes resulting in a `(3, 300, 300, 300)` array. So basically a lot of small matrices.

**Comment by hawkinsp** [comment]
It sounds very much as though the right implementation in Cusolver is the batched Jacobi one: https://docs.nvidia.com/cuda/cusolver/index.html#cuds-lt-t-gt-gesvdjbatch

I'll hook that up when I add batched SVD. Feel free to open a feature request to track.
